src/Discovery/src/Eureka/AppInfo/ApplicationInfoCollection.cs

```diff
@@ -92,79 +97,51 @@ internal void Add(ApplicationInfo app)
 
         foreach (InstanceInfo instance in app.Instances)
         {
-            AddInstanceToVip(instance);
+            AddToVipInstanceMap(instance);
         }
     }
 
-    private void AddInstanceToVip(InstanceInfo instance)
+    private void AddToVipInstanceMap(InstanceInfo instance)
     {
-        foreach (string vipAddress in ExpandVipAddresses(instance.VipAddress))
+        foreach (string vipAddress in ExpandVipAddresses(instance))
         {
-            AddInstanceToVip(instance, vipAddress, VipInstanceMap);
-        }
+            string addressUpper = vipAddress.ToUpperInvariant();
 
-        foreach (string secureVipAddress in ExpandVipAddresses(instance.SecureVipAddress))
-        {
-            AddInstanceToVip(instance, secureVipAddress, SecureVipInstanceMap);
+            ConcurrentDictionary<string, InstanceInfo> instancesById = VipInstanceMap.GetOrAdd(addressUpper, new ConcurrentDictionary<string, InstanceInfo>());
+            instancesById.AddOrUpdate(instance.InstanceId, _ => instance, (_, _) => instance);
         }
     }
 
-    private void AddInstanceToVip(InstanceInfo instance, string address, ConcurrentDictionary<string, ConcurrentDictionary<string, InstanceInfo>> dictionary)
+    private static HashSet<string> ExpandVipAddresses(InstanceInfo instance)
     {
-        lock (_addRemoveInstanceLock)
-        {
-            string addressUpper = address.ToUpperInvariant();
-
-            if (!dictionary.TryGetValue(addressUpper, out ConcurrentDictionary<string, InstanceInfo>? instances))
-            {
-                instances = new ConcurrentDictionary<string, InstanceInfo>();
-                dictionary[addressUpper] = instances;
-            }
+        HashSet<string> addresses = [];
 
-            instances[instance.InstanceId] = instance;
+        if (instance.SecureVipAddress != null)
+        {
+            string[] secureAddresses = instance.SecureVipAddress.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
+            addresses.UnionWith(secureAddresses);
```

Comment:
Is there an advantage of using `.UnionWith` here when we know `addresses` is empty, or is it more for alignment with line 128?